pkg/flow/internal/controller/node_config_import.go

```diff
@@ -0,0 +1,425 @@
+//
+// Run will immediately return ErrUnevaluated if Evaluate has never been called
+// successfully. Otherwise, Run will return nil.
+func (cn *ImportConfigNode) Run(ctx context.Context) error {
+	if cn.source == nil {
+		return ErrUnevaluated
+	}
+
+	newCtx, cancel := context.WithCancel(ctx)
+	defer cancel() // This will stop the children and the managed source.
+
+	errChan := make(chan error, 1)
+
+	runner := runner.New(func(node *ImportConfigNode) runner.Worker {
+		return &childRunner{
+			node:    node,
+			errChan: errChan,
+		}
+	})
+	defer runner.Stop()
+
+	updateTasks := func() error {
+		cn.mut.Lock()
+		defer cn.mut.Unlock()
+		cn.importChildrenRunning = true
+		var tasks []*ImportConfigNode
+		for _, value := range cn.importConfigNodesChildren {
+			tasks = append(tasks, value)
+		}
+
+		return runner.ApplyTasks(newCtx, tasks)
+	}
+
+	err := updateTasks()
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "import failed to run nested imports", "err", err)
+		cn.setRunHealth(component.HealthTypeExited, fmt.Sprintf("import shut down while trying to run nested imports with error: %s", err))
+		return err
+	}
+
+	go func() {
+		errChan <- cn.source.Run(newCtx)
+	}()
+
+	cn.setRunHealth(component.HealthTypeHealthy, "started import")
+	err = cn.run(errChan, updateTasks)
+
+	var exitMsg string
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "import exited with error", "err", err)
+		exitMsg = fmt.Sprintf("import shut down with error: %s", err)
+	} else {
+		level.Info(cn.logger).Log("msg", "import exited")
+		exitMsg = "import shut down normally"
+	}
+	cn.setRunHealth(component.HealthTypeExited, exitMsg)
+	return err
+}
+
+func (cn *ImportConfigNode) run(errChan chan error, updateTasks func() error) error {
+	for {
+		select {
+		case <-cn.importChildrenUpdateChan:
+			err := updateTasks()
+			if err != nil {
+				return err
+			}
+		case err := <-errChan:
+			return err
+		}
+	}
+}
+
+func (cn *ImportConfigNode) Label() string { return cn.label }
+
+// Block implements BlockNode and returns the current block of the managed config node.
+func (cn *ImportConfigNode) Block() *ast.BlockStmt { return cn.block }
+
+// NodeID implements dag.Node and returns the unique ID for the config node.
+func (cn *ImportConfigNode) NodeID() string { return cn.nodeID }
+
+// ImportedDeclares returns all declare blocks that it imported.
+func (cn *ImportConfigNode) ImportedDeclares() map[string]ast.Body {
+	cn.mut.RLock()
+	defer cn.mut.RUnlock()
+	return cn.importedDeclares
+}
+
+// ImportConfigNodesChildren returns the ImportConfigNodesChildren of this ImportConfigNode.
+func (cn *ImportConfigNode) ImportConfigNodesChildren() map[string]*ImportConfigNode {
+	cn.mut.Lock()
+	defer cn.mut.Unlock()
+	return cn.importConfigNodesChildren
+}
+
+type childRunner struct {
+	node    *ImportConfigNode
+	errChan chan error
+}
+
+func (cr *childRunner) Run(ctx context.Context) {
+	err := cr.node.Run(ctx)
+	if err != nil {
+		cr.errChan <- err
```

Comment:
I'm a little worried about the possibility of a goroutine leak here; if multiple children have fatal errors, then the third write will block forever here (the first write will be read in the `run` method; the second write will be written to the buffer; the third write will block because the buffer is full).

Do we need to be sending the failed signal back to the caller, or can we do something else here, such as logging errors or setting the health of the overall node?